src/zarr/codecs/pipeline.py

```diff
@@ -84,8 +84,13 @@ def from_dict(cls, data: Iterable[JSON | Codec], *, batch_size: int | None = Non
                 out.append(get_codec_class(name_parsed).from_dict(c))  # type: ignore[arg-type]
         return cls.from_list(out, batch_size=batch_size)
 
-    def to_dict(self) -> JSON:
-        return [c.to_dict() for c in self]
+    def to_dict(self) -> dict[str, JSON]:
+        return {
```

Comment:
Would it not be better to do `{'codecs': [c.to_dict() for c in self]}` here? It's a bit simpler, and matches the way the sharding codec stores multiple codecs.

Reply:
My thinking here was that the dict returned by `to_dict` should match the structure of the class itself, which has attributes like `array_array_codecs`, `array_bytes_codecs`, etc. But your comment made me notice two things:
- I didn't update the `from_dict` method, which assumes the old list representation
- The above mistake did not cause any tests to fail.

So I wonder if there is any need for `CodecPipeline` to implement `to_dict` and `from_dict` at all. [Despite my best efforts](https://github.com/zarr-developers/zarr-specs/issues/298), this data structure is not part of the zarr metadata, and we clearly are not using these serialization methods. I will experiment with removing inheritance from `Metadata` and see if anything breaks.